keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+- Documentations
+
+When a Security Profile release is shipped, all the manifest files are included in a .tar.gz file called security profile release bundle.
+The bundle can be hosted in github release page.
+
+The contents inside are self-contained:
+
+- Manifest files defining CRD and CRs of security profiles
+- Manifest files defining cluster-scope objects referenced by security profiles
+- Manifest files defining enforcers that are not expected to already be installed in every cluster, e.g. objects to deploy external admission controllers, and their required configurations
+
+To help tools to determine the right version of a security profile for a particular Kubernetes version,
+a version matrix is maintained in the source repo.
+The version matrix is a simple CSV file, defining:
+
+- Compatibility between a versioned security profile and Kubernetes versions
+- The security profile release bundle containing a versioned security profile
+- Deprecation of security profiles
+
+```
+default-1.0-0,1.9,std-security-profiles-0,DEPRECATED
+default-1.0-1,1.9,std-security-profiles-1,
+saas-multitenancy-1.0-1,1.9,std-security-profiles-1,
+default-1.0-1,1.10,std-security-profiles-1,
+saas-multitenancy-1.0-1,1.10,std-security-profiles-1,
+default-1.2-2,1.10,std-security-profiles-2,
+saas-multitenancy-1.2-2,1.10,std-security-profiles-2,
+```
+
+The columns are
+
+- security profile name and version
+- supported kubernetes version
+- security profile release bundle name
+- remarks, including DEPRECATED
+
+Each row records a single mapping between security profile, bundle, Kubernetes version, and remarks.
+
+#### Version Schema
+
+A security profile is versioned like `default-1.2-3`, where
+
+- `default-1` reflects the user scenarios this security profile covers
+- `.2` reflects the definitions of all rules and related objects
+- `-3` reflects the version of enforcement logic: admission controls, controllers, etc, mapping to version of release bundle
+
+Within a security profile release bundle, all resources defined in manifest files are labeled with bundle name. E.g.
+
+```
+profile.security.k8s.io/profile-bundle=std-security-profiles-3
+```
+
+#### Upgrade
+
+The general flow of upgrading to a new Security Profile is performed by:
+
+- Install a new version of security profile release bundle
+- Select a new version of security profile in `SecurityProfileSelector`
+- Uninstall the old security profile bundle
+
+Uninstall is easy as resources from a bundle are labeled with bundle name.
+With multiple bundles installed, it's possible multiple admission controls, controllers are running side-by-side.
+With help of labeled bundle name, the admission controls, controllers will stop enforcement if the currently selected security profile is not from the same bundle.
+With this mechanism, it's very simple and lightweight to switch back to the previous security profile if something unexpected happens with a new security profile.
```

Comment:
to the previous security profile version?

Reply:
not exactly the version. The user may upgrade from `default` to `saas`, and find out breaking workloads and decided to switch back to `default`.

Reply:
Makes sense